### ISSUE TITLE
Containment checks for separate types raises error

### ISSUE THREAD

**Issue by jsangmeister** [python/mypy#16089]
**Bug Report**

If a containment check is done on non-overlapping types, mypy reports an error although the code is type safe.

Further information: https://github.com/python/typing/discussions/1464

**To Reproduce**

```python
from enum import Enum

class Foo(str, Enum):
   BAR = "bar"

contains = "foo" in iter(Foo)
# or, even more simple but unusual code:
contains = "foo" in iter([1])
```

**Expected Behavior**

The code is type safe. The first example actually checks the containment in the iterator, the second example will always return false, but is still valid code. Mypy should not report an error for either case.

**Actual Behavior**

```
test.py:6: error: Unsupported operand types for in ("str" and "Iterator[Foo]")  [operator]
test.py:8: error: Unsupported operand types for in ("str" and "Iterator[int]")  [operator]
```

**Your Environment**

- Mypy version used: 1.5.1
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.11.3
